### DISCUSSION TITLE
git fetch error that I cannot fix via git stash

### DISCUSSION THREAD

**Discussion by Tillerz** [vladmandic/sdnext#926] [Q&A]
Hi,

I got the thing again that it complains about local changes. But I cannot fix it by using 'git stash' now. I have no idea what the problem is.

```
From https://github.com/vladmandic/automatic
 * branch              master     -> FETCH_HEAD
error: Your local changes to the following files would be overwritten by merge:
  extensions-builtin/sd-webui-controlnet
Please commit your changes or stash them before you merge.
Aborting
```

**Comment by vladmandic** [comment]
what version are you on? latest should attempt to fix that automatically if you run `webui --upgrade`

**Reply by Tillerz** [reply]
```
20:35:11-342071 ERROR    Error running git: . / pull origin master --autostash --rebase --force
20:35:11-344388 ERROR    Local changes detected: check setup.log for details
20:35:11-378958 INFO     Upgraded to version: a2485cf7 Fri May 12 21:12:24 2023 -0400
```

It actually doesn't upgrade to a2485cf7 because it gives this upgrade message on additional runs, too. At least for that one folder. :P

**Reply by evshiron** [reply]
What about `cd extensions-builtin/sd-webui-controlnet` and `git checkout .`? `extensions-builtin/sd-webui-controlnet` is a submodule which is another git repo.

**Reply by Tillerz** [reply]
`Updated 0 paths from the index`

**Reply by Tillerz** [reply]
Also I do not edit anything in there anywhere. I don't need to stash anything, the upgrade can just throw every change away if there is any that it has a problem with.

**Reply by vladmandic** [reply]
there is `webui --reset` which should do that.

**Reply by Tillerz** [reply]
`21:51:41-720454 ERROR    Error running git: None / merge --abort
21:51:44-102746 INFO     GIT reset complete`

hm. git version 2.40.1.windows.1 ... do I use a wrong git version maybe?

At least it's on `21:55:34-615396 INFO     Version: 819d105d Sat May 13 15:21:00 2023 -0400` now, and not complaining anymore.

**Reply by myndxero** [reply]
Getting the same problem, was about to post about it. I have not edited anything in the local controlnet repository. I do point the UI to custom folders for the models and such if that makes a difference.

My comp has been down over a week, just now getting it back up and running MAYBE. There's some issue with my video card be it power cable issues, psu issues, or failure.

**EDIT: I did a fresh install and the issue occurred there too.**

2023-05-13 22:09:16,947 | INFO | C:\automatic\installer.py | Starting SD.Next
2023-05-13 22:09:16,952 | INFO | C:\automatic\installer.py | Python 3.10.9 on Windows
2023-05-13 22:09:17,013 | DEBUG | C:\automatic\installer.py | Git 2.38.0.windows.1
2023-05-13 22:09:17,074 | INFO | C:\automatic\installer.py | Version: a2485cf7 Fri May 12 21:12:24 2023 -0400
2023-05-13 22:09:18,022 | DEBUG | C:\automatic\installer.py | Setting branch: . / master
2023-05-13 22:09:18,881 | ERROR | C:\automatic\installer.py | Error running git: . / pull origin master --autostash --rebase --force
2023-05-13 22:09:18,883 | ERROR | C:\automatic\installer.py | Local changes detected: check setup.log for details
2023-05-13 22:09:18,885 | DEBUG | C:\automatic\installer.py | Git output: Updating a2485cf7..926b8666

From https://github.com/vladmandic/automatic
 * branch              master     -> FETCH_HEAD
error: Your local changes to the following files would be overwritten by merge:
	extensions-builtin/sd-webui-controlnet
Please commit your changes or stash them before you merge.
Aborting

**Comment by vladmandic** [comment]
- delete setup.log
- update repo once manually using `git pull`
- run using `webui --upgrade`
- upload setup.log

**Comment by Tillerz** [comment]
And it's back. Did a manual git pull to see more. But it doesn't tell shit.

```automatic>git pull --verbose
POST git-upload-pack (198 bytes)
From https://github.com/vladmandic/automatic
 = [up to date]        master           -> origin/master
 = [up to date]        diffusers        -> origin/diffusers
 = [up to date]        directml-develop -> origin/directml-develop
 = [up to date]        directml-stable  -> origin/directml-stable
error: Your local changes to the following files would be overwritten by merge:
        extensions-builtin/sd-webui-controlnet
Please commit your changes or stash them before you merge.
Aborting
Updating 926b8666..c99c1410```

git stash does nothing here... would I have to do it on all subfolders in /sd-webui-controlnet, too? Can I tell git somehow to ignore all local change and overwrite, if needed?


EDIT: Just as a test I moved the whole folder extensions-builtin/ away before git pull. It did recreate extensions-builtin\sd-webui-controlnet

But shouldn't it NOT do that because extensions-builtin\sd-webui-controlnet is coming from a submodule?

**Reply by evshiron** [reply]
How about this:

```bash
cd automatic
rm -rf extensions-builtin/sd-webui-controlnet
git checkout .
git submodule --init --recusrive
```

**Reply by Tillerz** [reply]
git checkout . results in

`error: path 'extensions-builtin/sd-webui-controlnet' is unmerged`

**Reply by Tillerz** [reply]
Is the whole thing running better on linux? I can just install everything in a virtual environment. I think on linux you can also compile model files, does that speed up things? :)

**Reply by evshiron** [reply]
I am on Linux because ROCm only works on Linux for now. But I think Windows is totally fine if you are using a NVIDIA card.

The issue you have encountered seems a bit weird and is hard to diagnose remotely. Would you mind doing a fresh install in another directory and see if it works correctly?

**Comment by vladmandic** [comment]
if you run `webui --upgrade`  
and encounter `Error during repository upgrade`,
it should automatically attempt `Retrying repository upgrade` followed by `Running GIT reset`
that feature was added couple of days ago.

but...if you're on older version and you don't have that feature yet, to manually get out of bad situation you can run
```shell
git merge --abort
git fetch --all
git reset --hard origin/master
git checkout master
```
